doc/interoperability/pandas.rst

```diff
@@ -0,0 +1,299 @@
+of the issues with this are outlined below and sources (where available) of
+discussions and work-arounds (the latter is provided without guarantee that the
+workaround will still work) are provided. It should also be mentioned that if
+the DataFrame contains heterogenous data, the :meth:`.to_numpy()` property will
+create an in-memory copy of the DataFrame, thus using a NumPy array in the
+first place can be more memory efficient as well as avoiding some of the
+potential pitfalls when using DataFrames. If the DataFrame contains only
+homogenous data in the first place, no in-memory copy will be created using
+:meth:`.to_numpy`.
+
+Pandas in Pandas out
+====================
+
+It is currently not the case that all estimators in Scikit-learn return a
+DataFrame as an output when provided with a DataFrame as an input.
+
+Example for pandas working with Scikit-learn estimator/transformer:
+
+>>> import numpy as np
+>>> from sklearn.datasets import load_iris
+>>> import pandas as pd
+>>>
+>>> # make a dataframe
+>>> iris = load_iris()
+>>> X, y = iris.data[:-1,:], iris.target[:-1]
+>>> iris_pd = pd.DataFrame(X)
+>>> iris_pd.columns = iris.feature_names
+>>> iris_pd['target'] = y
+>>>
+>>> from sklearn.model_selection import train_test_split
+>>> train, test = train_test_split(iris_pd.copy(), test_size= 0.3)
+>>>
+>>> type(train)
+<class 'pandas.core.frame.DataFrame'>
+
+However, removing some random values from the dataset and using the
+:class:`~sklearn.impute.SimpleImputer` to replace the NaNs returns a NumPy
+array instead of a DataFrame even though we use a DataFrame as input.
+
+>>> rng = np.random.RandomState(42)
+>>> # selecting some random indices to replace
+>>> idx = train.index[rng.binomial(1, 0.2, train.shape[0]).astype(bool)]
+>>> train.loc[idx, 'sepal length (cm)'] = np.nan
+>>>
+>>> from sklearn.impute import SimpleImputer
+>>>
+>>> imputer = SimpleImputer()
+>>> X = imputer.fit_transform(train)
+>>> type(X)
+<class 'numpy.ndarray'>
+
+As this example shows, at the moment it is not guaranteed that Scikit-learn
+operators with :meth:`.fit`, :meth:`.transform` (and :meth:`.predict`)
+capability support pandas in pandas out. However, there are ways around this
+such as an example given
+`here <https://github.com/scikit-learn/scikit-learn/issues/5523#issuecomment-171674105>`__
+show, where adding additional functionality to the StandardScaler class adds
+the pandas in pandas out capability. Care should be taken as this does not
+take care of the column ordering problem that is discussed in the next section.
+
+The column ordering problem
+===========================
+
+Because Scikit-learn transforms DataFrames to NumPy arrays, it should be
+assumed, that all information and benefits of column names is lost and that
+from that point forward, only column order and not column labels stay relevant.
+This can cause problems when e.g. pickling a trained estimator and later
```

Comment:
this is a problem in general with using new data for predict, not only when having pickled a trained estimator, so I would just scope this problem as "predicting on unseen data" where the new data have a different column order.